### PR TITLE
allocator: Re-use randomly generated ID sequence between allocations

### PR DESCRIPTION
**Pull request by tgraf**
While benchmarking the remaining cost of allocations, math/rand usage is the
clear bottleneck:
```
    1560ms 12.80% 25.92%     4830ms 39.62%  math/rand.(*Rand).Perm /usr/local/go/src/math/rand/rand.go
    1440ms 11.81% 37.74%     2510ms 20.59%  math/rand.(*Rand).Int31n /usr/local/go/src/math/rand/rand.go
    1140ms  9.35% 47.09%     1590ms 13.04%  runtime.scanobject /usr/local/go/src/runtime/mgcmark.go
     540ms  4.43% 51.52%      540ms  4.43%  math/rand.(*rngSource).Int63 /usr/local/go/src/math/rand/rng.go
     510ms  4.18% 55.70%      510ms  4.18%  runtime.futex /usr/local/go/src/runtime/sys_linux_amd64.s
     460ms  3.77% 59.47%     2970ms 24.36%  math/rand.(*Rand).Intn /usr/local/go/src/math/rand/rand.go
     280ms  2.30% 61.77%     1070ms  8.78%  math/rand.(*Rand).Int31 /usr/local/go/src/math/rand/rand.go
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4573)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

File: pkg/kvstore/allocator/allocator.go [hunk 1]
```diff
@@ -380,21 +383,32 @@ var (
...
+				return id, id.String()
+			}
+		}
 
-	for _, r := range idRandomizer.Perm(int(a.max - a.min + 1)) {
-		id := ID(r) + a.min
-		tried++
-		if _, ok := a.cache[id]; !ok && a.localKeys.lookupID(id) == "" {
-			return id, id.String()
+		if attempt == 0 {
```
Comment (rlenglet) [inline pkg/kvstore/allocator/allocator.go:419]: We're doing 2 attempts here. If the 2nd attempt fails, this method returns leaving `a.randomIDs` empty. So in the next call to `selectAvailableID()`, the first attempt will iterate over an empty slice and do nothing, so only the 2nd attempt will do anything.
I would replace this block with a check before the 2nd for loop above:
```
if len(a.randomIDs) == 0 {
    a.newRandomIDs()
}
for i, r := range a.randomIDs {
...
}
```

File: pkg/kvstore/allocator/allocator.go [hunk 2]
```diff
@@ -380,21 +383,32 @@ var (
...
 	defer a.mutex.RUnlock()
 
-	tried := 0
+	for attempt := 0; attempt < 2; attempt++ {
+		for i, r := range a.randomIDs {
+			id := ID(r) + a.min
+			if _, ok := a.cache[id]; !ok && a.localKeys.lookupID(id) == "" {
+				// remove the previously tried IDs that are already in
+				// use from the list of IDs to attempt allocation
+				if len(a.randomIDs) > 0 {
```
Comment (rlenglet) [inline pkg/kvstore/allocator/allocator.go]: Remove this test. This is always true since we're currently iterating over that slice.

**Review by rlenglet — DISMISSED** [review]

File: pkg/kvstore/allocator/allocator.go (hunk 1 above)
Comment (tgraf) [reply pkg/kvstore/allocator/allocator.go:419]: That's intentional. If we start out with an empty slice, the 2nd attempt covers a random sequence over the entire available ID space. I can leave a comment if that is not clear.

Comment (tgraf) [reply pkg/kvstore/allocator/allocator.go:419]: Added comment

**Comment by tgraf** [comment]
test-me-please

**Review by rlenglet — APPROVED** [review]
That's more clear, thanks!

**Pull request merged**